File: wlr/OGUR/OGUR/Creatures/Minions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using OGUR.Skills;
using SPX.Sprites;
using OGUR.Strategies;
using OGUR.Entities;
using SPX.Entities;
using SPX.Core;

namespace OGUR.Creatures
{
    class Minion : ICreature
    {
        public Minion(int actorType = OgurActorType.MINION,float coolDown = Stats.DefaultCoolDown)
        {
            m_actorType = actorType;
            m_baseStats = new Stats(80f,999f,0f,0f,0f,0f,0f,0f,0f,coolDown);
        }
        public void Init(ICreature source,SkillEffect effectGraphic)
        {
            m_master = source;
            Setup(source.GetLocation(), m_actorType, m_baseStats,null,false);
            if (null != effectGraphic)
            {
                SetSkillVector(effectGraphic.GetDirection().Rotate180());
            }
            else
            {
                SetSkillVector(new Point2(0, 1));
            }
            m_strategy.AddTargets(m_master);
        }
        protected void Add(string skill)
        {
            if (m_skills == null)
            {
                m_skills = new SkillPool(this);
            }
            m_skills.Add(skill);
        }
    }

    class AcidNozzle : Minion
    {
        public AcidNozzle()
            : base(OgurActorType.MINION,50f)
        {
            Add(SkillId.ACID_DRIP);
            m_composition.Add(Elements.EARTH);
            m_strategy = new MinionRotate(this);
        }
    }

    class DartTrap : Minion
    {
        public DartTrap()
            : base(OgurActorType.MINION)
        {
            m_strategy = new MinionFire(this);
            Add(SkillId.DART);
            m_composition.Add(Elements.DARK);
        }
    }

    class Explosion : Minion
    {
        public Explosion()
            : base(OgurActorType.MINION)
        {
            m_strategy = new MinionExplode(this);
            Add(SkillId.EXPLODE);
            m_composition.Add(Elements.FIRE);
        }
    }

    class VaporCloud : Minion
    {
        private ICreature m_host = null;
        public VaporCloud()
            : base(OgurActorType.MINION)
        {
            m_strategy = new MinionCloud(this);
            Add(SkillId.VAPOR_CLOUD);
            m_composition.Add(Elements.WATER);
        }
        public override void Update()
        {
            base.Update();
            if (m_host == null)
            {
                foreach (var creature in EntityManager.GetActorsAt(m_location))
                {
                    if (creature != this)
                    {
                        m_host = creature as ICreature;
                    }
                }
                if (m_host == null)
                {
                    SetInactive();
                }
            }
            if (m_host != null)
            {
                SetLocation(m_host.GetLocation());
            }
        }
    }
}

File: port-workspace/aigilas-ported/src/com/aigilas/creatures/Minions.java
package com.aigilas.creatures;import com.xna.wrapper.*;import java.util.*;import com.aigilas.skills.*;import com.spx.sprites.*;import com.aigilas.strategies.*;import com.aigilas.entities.*;import com.spx.entities.*;import com.spx.core.*;
    public class Minion  extends  ICreature
    {
        public Minion(int actorType = AigilasActorType.MINION,float coolDown = Stats.DefaultCoolDown)
        {
            _actorType = actorType;
            _baseStats = new Stats(80f,999f,0f,0f,0f,0f,0f,0f,0f,coolDown);
        }
        public void Init(ICreature source,SkillEffect effectGraphic)
        {
            _master = source;
            Setup(source.GetLocation(), _actorType, _baseStats,null,false);
            if (null != effectGraphic)
            {
                SetSkillVector(effectGraphic.GetDirection().Rotate180());
            }
            else
            {
                SetSkillVector(new Point2(0, 1));
            }
            _strategy.addTargets(_master);
        }
        protected void Add(String skill)
        {
            if (_skills == null)
            {
                _skills = new SkillPool(this);
            }
            _skills.add(skill);
        }
    }

    class AcidNozzle  extends  Minion
    {
        public AcidNozzle()
            { super(AigilasActorType.MINION,50f);            Add(SkillId.ACID_DRIP);
            _composition.add(Elements.EARTH);
            _strategy = StrategyFactory.Create(Strategy.MinionRotate,this);
        }
    }

    class DartTrap  extends  Minion
    {
        public DartTrap()
            { super(AigilasActorType.MINION);            _strategy = StrategyFactory.Create(Strategy.MinionFire, this);
            Add(SkillId.DART);
            _composition.add(Elements.DARK);
        }
    }

    class Explosion  extends  Minion
    {
        public Explosion()
            { super(AigilasActorType.MINION);            _strategy = StrategyFactory.Create(Strategy.MinionOneUse, this);
            Add(SkillId.EXPLODE);
            _composition.add(Elements.FIRE);
        }
    }

    public class IceShard  extends  Minion
    {
        public IceShard()
            { super(AigilasActorType.MINION);            _strategy = StrategyFactory.Create(Strategy.MinionOneUse, this);
            Add(SkillId.ICE_SHARD);
            _composition.add(Elements.WATER);
        }
    }

    class PoisonCloud  extends  Minion
    {
        public PoisonCloud()
            { super(AigilasActorType.MINION);            _strategy = StrategyFactory.Create(Strategy.MinionOneUse, this);
            Add(SkillId.POISON_CLOUD);
            _composition.add(Elements.MENTAL);
        }
    }

    class VaporCloud  extends  Minion
    {
        private ICreature _host = null;
        public VaporCloud()
            { super(AigilasActorType.MINION);            _strategy = StrategyFactory.Create(Strategy.MinionCloud, this);
            Add(SkillId.VAPOR_CLOUD);
            _composition.add(Elements.WATER);
        }
@Override        public  void Update()
        {
            super.Update();
            if (_host == null)
            {
                for (var creature:EntityManager.GetActorsAt(_location))
                {
                    if (creature != this)
                    {
                        _host = creature as ICreature;
                    }
                }
                if (_host == null)
                {
                    SetInactive();
                }
            }
            if (_host != null)
            {
                SetLocation(_host.GetLocation());
            }
        }
    }
